### ISSUE TITLE
Notation priority is unclear

### ISSUE THREAD

**Issue by yannl35133** [Deducteam/lambdapi#653]
In the documentation, an instinctive definition for priority is given.

However, no mention is given for the range of values that they can take, and I discovered that the lexer doesn't accept strictly negative values, nor does the parser tolerate the value zero 
```lambdapi
constant symbol Prop : TYPE;

constant symbol ⇒ : Prop → Prop → Prop;
notation ⇒ infix right 0;

symbol s (a: Prop) ≔ a ⇒ a;
```
`[/path/test.lp, 6:29-30] Operator conflict between "a" and "⇒"`

Can you provide more explicit documentation?
I also wonder how standard application interacts with this priority system. (My testing indicates that it has maximum priority, please confirm)

**Comment by gabrielhdt** [comment]
Hi, I confirm: application has highest priority. Operators, priorities and associativity are handled using the Pratt parsing algorithm, the source code can be found there https://forge.tedomum.net/koizel/pratter, and you may look at the examples in https://forge.tedomum.net/koizel/pratter/-/blob/master/t/main.ml to have a clear idea of the behaviour. It's true though, that the documentation could be more precise.

**Comment by fblanqui** [comment]
Hello @gabrielhdt .

Could you please make a PR fixing the doc?

I am also wondering how infix and prefix symbol applications interact. For instance, does the parsing of "-x+-y" change with the priorities of + and -? It would be nice to give more details about that in the documentation as well.

Concerning prefix symbols, we should have --x parsed as -(-x) if - is declared prefix. Is it tested? In this case, is it still correct to say: "application has highest priority"?

By the way, why not allowing postfix symbols as well (like "!" for factorial)?

**Comment by gabrielhdt** [comment]
Hi,

> Could you please make a PR fixing the doc?

I think it should rather be in the documentation of the pratter library, after all, it's its job, but anyway I have to document it, indeed.

> Concerning prefix symbols, we should have --x parsed as -(-x) if - is declared prefix. Is it tested? In this case, is it still correct to say: "application has highest priority"?

I don't think it's tested, but it must be the case. I'll add a test.

> By the way, why not allowing postfix symbols as well (like "!" for factorial)?

There is no reason for not allowing postfix other than "it hasn't been done". We can even implement ternary operators such as `if . then . else .`.

**Comment by fblanqui** [comment]
Le 03/05/2021 à 08:06, Gabriel Hondet a écrit :
>
> I think it should rather be in the documentation of the pratter 
> library, after all, it's its job
>
Not only because LP users do not need to know that LP uses Pratter.

> , but anyway I have to document it, indeed.
>
Thanks.

**Comment by gabrielhdt** [comment]
> I am also wondering how infix and prefix symbol applications interact. For instance, does the parsing of "-x+-y" change with the priorities of + and -? It would be nice to give more details about that in the documentation as well.

Parsing changes indeed: `!x + y = (!x) + y` if `!` has a higher binding power than `+`, and `!x + y = !(x + y)` otherwise.

> Concerning prefix symbols, we should have --x parsed as -(-x) if - is declared prefix. Is it tested? In this case, is it still correct to say: "application has highest priority"?

It is the case.

**Comment by fblanqui** [comment]
Ok. But what do you mean exactly by "application has highest priority" because all these expressions are just application trees with leaves in different order?

**Comment by gabrielhdt** [comment]
I mean regular application of symbols rather than application of prefix argument: we have `!f x = !(f x)`, whatever the binding power of `!` is.

**Comment by fblanqui** [comment]
This should be fixed in #664.

**Comment by yannl35133** [comment]
The negative priorities are also now accepted by the Pratt parser v1.2, I see no reason to keep this open

**Comment by fblanqui** [comment]
Note that issues should be closed only after merge.

**Comment by fblanqui** [comment]
Fixed in #664.
